### PR TITLE
[BugFix] Fix not enough hosts error when BE is decommissioned

### LINKED ISSUE #6145: Can not create table when one BE is decommissioned
### Steps to reproduce the behavior (Required)
image

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by nshangyiming**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [x] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #6145 

## Problem Summary(Required) ：

1. When multiple BEs reside on the same host, only one BE will be chosen when create table. But if some of them are decommissioned,  FE still might choose the decommissioned one, this will cause the problem described in the issue.
2. Also refine some codes based on sonar code smell.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 9 / 20 (45.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|1|00.00%|[549]|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|1|00.00%|[846]|
|:large_blue_circle:|com/starrocks/http/rest/LoadAction.java|0|1|00.00%|[80]|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|6|14|42.86%|[296, 409, 627, 831, 832, 833, 835, 985]|
|:large_blue_circle:|com/starrocks/http/rest/TransactionLoadAction.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 12 / 22 (54.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|1|00.00%|[549]|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|1|00.00%|[846]|
|:large_blue_circle:|com/starrocks/http/rest/LoadAction.java|0|1|00.00%|[80]|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|7|14|50.00%|[296, 409, 831, 832, 833, 835, 985]|
|:large_blue_circle:|com/starrocks/http/rest/TransactionLoadAction.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/system/ComputeNode.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 12 / 22 (54.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|1|00.00%|[549]|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|1|00.00%|[846]|
|:large_blue_circle:|com/starrocks/http/rest/LoadAction.java|0|1|00.00%|[80]|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|7|14|50.00%|[296, 409, 831, 832, 833, 835, 985]|
|:large_blue_circle:|com/starrocks/http/rest/TransactionLoadAction.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/system/ComputeNode.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java [hunk 1]
```diff
@@ -592,24 +592,24 @@ public List<Backend> getBackends() {
...
                 getBackends().stream().filter(v -> !v.diskExceedLimit()).collect(Collectors.toList());
-        return seqChooseBackendIds(backendNum, needAlive, isCreate, backends);
+        return seqChooseBackendIds(backendNum, needAvailable, isCreate, backends);
     }
 
     // choose backends by round-robin
     // return null if not enough backend
     // use synchronized to run serially
-    public synchronized List<Long> seqChooseBackendIds(int backendNum, boolean needAlive, boolean isCreate,
+    public synchronized List<Long> seqChooseBackendIds(int backendNum, boolean needAvailable, boolean isCreate,
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java:612]: If you change the needAlive to needAvailable, you shoud change all the Alive to the Available

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java [hunk 2]
```diff
@@ -843,7 +843,7 @@ private Partition resetPartitionForRestore(OlapTable localTbl, OlapTable remoteT
                 List<Long> beIds =
                         GlobalStateMgr.getCurrentSystemInfo().seqChooseBackendIds(restoreReplicationNum, true,
                                 true);
-                if (beIds == null) {
+                if (beIds.isEmpty()) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:846]: `CollectionUtils.isEmpty` is more safe? To check null or empty?

File: fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java [hunk 3]
```diff
@@ -697,10 +696,10 @@ public synchronized List<Long> seqChooseBackendIds(int backendNum, boolean needA
 
         // debug
         for (Backend backend : backends) {
-            LOG.debug("random select: {}", backend.toString());
+            LOG.debug("random select: {}", backend);
         }
 
-        return null;
+        return new ArrayList<>();
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/system/SystemInfoService.java:702]: `Collections.emptyList()` ?

**Review by murphyatwork — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java (hunk 2 above)
Comment (nshangyiming) [reply fe/fe-core/src/main/java/com/starrocks/backup/RestoreJob.java:846]: Currently, it won't return null, but anyway using `CollectionUtils.isEmpty` looks like a better practice. I will refine it.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 12 / 23 (52.17%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/Utils.java|0|1|00.00%|[35]|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|0|1|00.00%|[551]|
|:large_blue_circle:|com/starrocks/backup/RestoreJob.java|0|1|00.00%|[847]|
|:large_blue_circle:|com/starrocks/http/rest/LoadAction.java|0|1|00.00%|[81]|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|7|14|50.00%|[296, 409, 831, 832, 833, 835, 985]|
|:large_blue_circle:|com/starrocks/http/rest/TransactionLoadAction.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/system/ComputeNode.java|2|2|100.00%|[]|

**Comment by nshangyiming** [comment]
@mofeiatwork need another + 1

**Review by gengjun-git — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**